common/version.tmpl.cpp

```diff
@@ -0,0 +1,51 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "common/version.h"
+
+#include <string_view>
+
+namespace Carbon {
+
+// A simplistic string-to-integer routine that is consteval for compile-time
+// extracting specific components of the version from the string form. We use
+// `std::string_view` for its broader `constexpr` API.
+static consteval auto ToInt(std::string_view str) -> int {
+  int result = 0;
+  for (;;) {
+    result += str.front() - '0';
+    str.remove_prefix(1);
+    if (str.empty()) {
+      break;
+    }
+    result *= 10;
+  }
+  return result;
+}
+
+static consteval auto MajorVersion(std::string_view str) -> int {
+  return ToInt(str.substr(0, str.find('.')));
+}
+
+static consteval auto MinorVersion(std::string_view str) -> int {
+  str.remove_prefix(str.find('.') + 1);
+  return ToInt(str.substr(0, str.find('.')));
+}
+
+static consteval auto PatchVersion(std::string_view str) -> int {
+  str.remove_prefix(str.find('.') + 1);
+  str.remove_prefix(str.find('.') + 1);
+  // Note that searching for `-` may find the end of the string if there is no
+  // pre-release component, but that produces the correct result here.
+  return ToInt(str.substr(0, str.find('-')));
+}
+
+// The major, minor, and patch versions are always provided and stable. They
+// don't depend on build stamping or introduce caching issues. Provide normal
+// strong definitions.
+constexpr int Version::Major = MajorVersion("$VERSION");
+constexpr int Version::Minor = MinorVersion("$VERSION");
+constexpr int Version::Patch = PatchVersion("$VERSION");
```

Comment:
Yeah, I went back and forth on this. I can go back, I'm of two minds.

Basically, we need both the string form and integer form in C++. We only need the string form in Bazel. So we can either keep the integers all the way from Bazel into here, and we get less parsing of the version string, but more complex logic in Bazel that isn't needed on the Bazel side, more different things are expanded in the template, and it's quite opaque in the C++ code that the integers and the string match exactly.

This way, we have less logic in Bazel, fewer expanded template things, and its obvious the connection. But we have to write the constexpr parsing.

For me, the tie break was that using integer template expansions here makes the syntax trivially broken, and so editting this file is really painful. Whereas here, we can constrain the templated bits to be in string literals (or a weird attribute), and it keeps the syntax checking working which is somewhat nice.

But that's a pretty minor point. WDYT?

Reply:
What do you think of changing version_base.bzl to look more like:

```
version_major = '0'
version_minor = '0'
version_patch = '0'
```

That way Bazel gets the string form with `"{0}.{1}.{2}".format(version_major, version_minor, version_patch)`, no parsing?

Reply:
As I said, that was one of the things I tried. =]

The downside is that Bazel only needs the string form, so the integers will only be used for creating the substitutions for the C++ code, and there will be more substitutions there.

It removes the C++ parsing, but I'd feel tempted to keep that code in a static_assert to document that these aren't independent and the string and number match. But at that point, having them broken out in Bazel seems a bit weird.

What would help for me is to understand why this seems better? Is it getting ride of the constexpr parsing code? something else?

Reply:
I would think it's better because it's less code overall... It gets rid of the constexpr parsing code, and replaces it with a single format string. It's also clearer in version_base that these are integers, not an arbitrary string (i.e., adding more digits "0.0.0.1" will more clearly not work)

Reply:
Note, I'm not sure what you'd static assert on. Just don't pass the complete string to C++ at all. Have a function that does the format on request.

Reply:
I'm happy to use three integers in the `version_base.bzl` file if that seems cleaner. No strong opinion there, let me know if that would be an improvement in isolation.

Beyond that:

I think passing the integers as such into C++ still seems like the worse tradeoff to me. It's not a tiny amount of code to set up all the substitutions in starlark. And for me, a single line of starlark is more expensive than several lines of C++ code.

But you can't avoid passing the string to C++ I think... Or at least, I think that would be more complex to try to achieve. You'd need to pass:
- The 3 version numbers
- What kind of pre-release it is (if any)
- A string pre-release suffix (or both a string and a number suffix depending on the kind)

And it's really important that these are assembled into a string using the exact same format in both Starlark and C++. It seemed better to assemble all of these into a string in one place (and that has to be Starlark), and disassemble the components in another. And ideally, to not have the once-decomposed components travel across a really opaque boundary like code generation between starlark and C++.

Reply:
Note, approving right now, but I still think this is something that could maybe be reduced further... I'm maybe more okay with a few boilerplate lines passing values, versus a volume of parsing/validation code. But can address that separately.

Reply:
Thanks, and happy to talk more about simplifying here and what the options / tradeoffs are.